boot/freeldr/freeldr/arch/realmode/helpers_pc98.inc

```diff
@@ -80,15 +80,22 @@ writechr:
     pushf
     pusha
 
+    mov ax, word ptr VramSegment
```

Comment:
There is a nice `LES` instruction you can use. This suggestion has been tested to work correctly:
```writechr:
    pushf
    pusha

    /* Check if the VRAM segment was initialized */
    mov dx, word ptr ds:[VramSegment]
    test dx, dx
    jnz .writechr_write

    /* High-resolution mode check */
    test byte ptr ds:[HEX(501)], HEX(08)
    mov dx, HEX(0A000)
    jz .writechr_test_done
    mov dh, HEX(0E)

.writechr_test_done:
    mov word ptr ds:[VramSegment], dx

.writechr_write:
    /* Load the VRAM segment and offset */
    les di, dword ptr ds:[VramOffset]

    /* Two bytes per character */
    xor ah, ah

    /* Write ASCII directly to the VRAM */
    stosw

    /* Update the start position */
    mov word ptr ds:[VramOffset], di

    popa
    popf
    ret

VramOffset:
    .word 0
VramSegment:
    .word 0